pkg/kclient/all.go

```diff
@@ -26,41 +26,47 @@ func (c *Client) GetAllResourcesFromSelector(selector string, ns string) ([]unst
 }
 
 func getAllResources(client dynamic.Interface, apis []apiResource, ns string, selector string) ([]unstructured.Unstructured, error) {
-	var mu sync.Mutex
-	var wg sync.WaitGroup
 	var out []unstructured.Unstructured
+	outChan := make(chan []unstructured.Unstructured)
 
-	start := time.Now()
-	klog.V(2).Infof("starting to query %d APIs in concurrently", len(apis))
-
-	var errResult error
+	var apisOfInterest []apiResource
 	for _, api := range apis {
 		if !api.r.Namespaced {
 			klog.V(4).Infof("[query api] api (%s) is non-namespaced, skipping", api.r.Name)
 			continue
 		}
-		wg.Add(1)
-		go func(a apiResource) {
-			defer wg.Done()
-			klog.V(4).Infof("[query api] start: %s", a.GroupVersionResource())
-			v, err := queryAPI(client, a, ns, selector)
+		apisOfInterest = append(apisOfInterest, api)
+	}
+
+	start := time.Now()
+	group := new(errgroup.Group) // an error group errors when any of the go routines encounters an error
+	klog.V(2).Infof("starting to concurrently query %d APIs", len(apis))
+
+	for _, api := range apisOfInterest {
+		api := api // shadowing because go vet complains "loop variable api captured by func literal"
+		group.Go(func() error {
+			klog.V(4).Infof("[query api] start: %s", api.GroupVersionResource())
+			v, err := queryAPI(client, api, ns, selector)
 			if err != nil {
-				klog.V(4).Infof("[query api] error querying: %s, error=%v", a.GroupVersionResource(), err)
-				errResult = err
-				return
+				klog.V(4).Infof("[query api] error querying: %s, error=%v", api.GroupVersionResource(), err)
+				return err
 			}
-			mu.Lock()
-			out = append(out, v...)
-			mu.Unlock()
-			klog.V(4).Infof("[query api]  done: %s, found %d apis", a.GroupVersionResource(), len(v))
-		}(api)
+			outChan <- v
+			klog.V(4).Infof("[query api]  done: %s, found %d apis", api.GroupVersionResource(), len(v))
+			return nil
+		})
 	}
 
+	go func() {
+		for val := range outChan {
```

Comment:
I think this goroutine will leak and keep running indefinitely.
Maybe I'm wrong, but from my understanding, this `for-range` loop will finish only when the channel is closed.
So if we run a long-running process like `odo logs --follow`, this goroutine will stay here forever, waiting for `outChan` to be closed.
Also, I noticed that `getAllResources` might be called several times (tested with a component in both Dev and Deploy modes), and this led to multiple goroutines leaked and waiting forever for `outChan` to be closed.
Maybe it makes more sense to have a dedicated goroutine that calls `group.Wait()` and closes `outChan`? Then outside, this `for-range` loop. This way, we are sure that `outChan` is closed properly after all goroutines fired have returned (successfully or not), so no risk of panic'ing, IMO. I guess this might require to capture the error returned by `group.Wait()` in a dedicated error channel..

Maybe it is more clear with some code :)
```go
klog.V(2).Infof("fired up all goroutines to query APIs")

// Buffered channel of size 1 so as not to block when writing the error,
// but `group.Wait()` waits for the goroutines in the group to return (or any of them returning an error).
errChan := make(chan error, 1)
go func() {
  errChan <- group.Wait()
  close(outChan)
})

for v := range outChan {
  out = append(out, v...)
}

//At this point, we are sure that outChan is closed and we have everything,
// even if any of the goroutines fired returned an error.

klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
klog.V(2).Infof("query result: objects=%d", len(out))
return out, <-errChan
```
 
What do you think?

Reply:
> Maybe I'm wrong, but from my understanding, this `for-range` loop will finish only when the channel is closed.

Nope, you are right. I wasn't aware of this till I read your comment and looked it up. 



> ```go
> // Buffered channel of size 1 so as not to block when writing the error,
> // but `group.Wait()` waits for the goroutines in the group to return
> // (or any of them returning an error).
> ```

I don't understand the rationale behind buffer size of 1. Based on the [documentation of `Wait()`](https://pkg.go.dev/golang.org/x/sync/errgroup#Group.Wait), it's going to return the first non-nil error if any of the go routines had any. I don't think the channel's going to be written to again by `group.Wait()`.

Otherwise, what you have suggested, makes sense overall.

Reply:
[b3001cb](https://github.com/redhat-developer/odo/pull/5973/commits/b3001cbdedbcb1aa354f166ebdc7bd0a7af262d0)

Reply:
> > Maybe I'm wrong, but from my understanding, this `for-range` loop will finish only when the channel is closed.
> 
> Nope, you are right. I wasn't aware of this till I read your comment and looked it up.
> 
> > ```go
> > // Buffered channel of size 1 so as not to block when writing the error,
> > // but `group.Wait()` waits for the goroutines in the group to return
> > // (or any of them returning an error).
> > ```
> 
> I don't understand the rationale behind buffer size of 1. Based on the [documentation of `Wait()`](https://pkg.go.dev/golang.org/x/sync/errgroup#Group.Wait), it's going to return the first non-nil error if any of the go routines had any. I don't think the channel's going to be written to again by `group.Wait()`.
> 

This was mainly to have a non-blocking write to `errChan`. Otherwise, this write operation waits for a read, which actually happens after the `for-range` loop (which also blocks until `outChan` is closed).

Reply:
> Otherwise, this write operation waits for a read, which only happens after the `for-range` loop (which also blocks until `outChan` is closed).

Your concern is valid. But when a go routine that is waiting for something for longer time is encountered, the go scheduler takes care of being fair and not letting it run all the time. This is something I learned from [this talk](https://www.youtube.com/watch?v=wQpC99Xu1U4). I watched it many months back when I had not even started writing much concurrent code. So if you find something different from my understanding, let me know!

Reply:
> > Otherwise, this write operation waits for a read, which only happens after the `for-range` loop (which also blocks until `outChan` is closed).
> 
> Your concern is valid. But when a go routine that is waiting for something for longer time is encountered, the go scheduler takes care of being fair and not letting it run all the time. This is something I learned from [this talk](https://www.youtube.com/watch?v=wQpC99Xu1U4). I watched it many months back when I had not even started writing much concurrent code. So if you find something different from my understanding, let me know!

> This was mainly to have a non-blocking write to `errChan`.

Even with the above concern, I can't understand how a buffer size of 1 would have helped. If the `range` block was going to block, what difference was making the `errChan` buffered going to make? 🤔

Reply:
> > > Otherwise, this write operation waits for a read, which only happens after the `for-range` loop (which also blocks until `outChan` is closed).
> 
> Your concern is valid. But when a go routine that is waiting for something for longer time is encountered, the go scheduler takes care of being fair and not letting it run all the time. This is something I learned from [this talk](https://www.youtube.com/watch?v=wQpC99Xu1U4). I watched it many months back when I had not even started writing much concurrent code. So if you find something different from my understanding, let me know!
> 
Interesting - thanks for the link to this talk. Will watch it soon.

> > This was mainly to have a non-blocking write to `errChan`.
> 
> Even with the above concern, I can't understand how a buffer size of 1 would have helped. If the `range` block was going to block, what difference was making the `errChan` buffered going to make? thinking
>
In my code sample, the buffer size of 1 was there mainly to be able to call `close(outChan)`. Otherwise, `errChan <- group.Wait()` would block waiting for a reader; but in my example, we read from `errChan` only after the `for-range` loop (which itself blocks until `close(outChan)` is called). So deadlock without the buffer size of 1.